### ISSUE TITLE
Provide option for fixtures to not be shared between test methods

### ISSUE THREAD

**Issue by chris-smith-zocdoc** [nunit/nunit#2105]
This was originally brought up in https://github.com/nunit/nunit/issues/164#issuecomment-272480602 but I'm submitting a new issue now that #164 is closed

**Expected Behavior**
`[SetUp]` should run before every `[Test]` in a thread safe way, allowing the usage of `[Parallelizable(ParallelScope.All)]` to parallelize test methods.  

**Observed Behavior**
Test methods executing in parallel operate on the same instance of the test class, this allows
`[SetUp]` (or other state modifying actions) to be called from another thread during the execution of another test, interfering with the currently executing test by modifying state in the test class.

**Contrived Example**
```cs
 [Parallelizable(ParallelScope.All)]
 class ParallelTests
 {
     private int count;

     [SetUp]
     public void Setup()
     {
         count = 0;
         Thread.Sleep(TimeSpan.FromSeconds(4));
         count++;
     }

     [Test]
     public void Test1()
     {
         Assert.AreEqual(1, count);
     }

     [Test]
     public void Test2()
     {
         Assert.AreEqual(1, count);
     }
 }
```

These tests will fail with "Expected 1 but was 2". Using a different ParallelScope these tests pass.

**Proposal**
Each test should have its own instance of the test class, this will make it significantly easier to for developers to write test methods that execute in parallel. 

If we want `[OneTimeSetUp]` to work with this proposal, it would most likely be a requirement for OneTimeSetUp methods to be static so they can operate on data shared across all the test classes.

**Comment by CharliePoole** [comment]
@chris-smith-zocdoc As I wrote originally, I'm not in favor of this - which is why it wasn't done as part of #164.

As you indicate, having individual tests affect the object state limits your ability to run tests in parallel. It also tends to make your tests interdependent, which has traditionally been considered a user error in the construction of tests. You correctly labelled your example as "contrived" - one might even say "pathological" or "erroneous" depending on the viewpoint.

Traditionally, the NUnit view has been that we can use the same instance for multiple tests, as well as instance methods for tests, setup and teardown, __because__ we trust the user not to change state - or only to do it in a very controlled way, usually in SetUp and TearDown.

Others have said "But you should make it __impossible__ for users to make those errors." Up to now we have not agreed that that is our role. In very rare cases, there are good reasons to change state. NUnit is a very sharp tool - we want it to be so - and it has to be used with care.

OTOH, I can see arguments for using per-test instances in some situations. We have discussed doing that as an option or as part of a new kind of TestFixture, although I don't think it has been documented as an issue. However, I would see this as entirely orthogonal to parallel test case execution, as most NUnit features are orthogonal by design

In most cases, tests themselves do not modify object state. Where they do, the intent of the current design is that users should specify parallelization at the Fixture level. I realize that this assumes a certain level of skill on the part of those writing tests. It works against organizations that assign test-writing as a separate activity to less experienced programmers. To be honest, I"m happy with that consequence. 😈 

I'll leave this issue as "new" so others can comment as well.

**Comment by tparikka** [comment]
@CharliePoole I am one of those less experienced programmers writing tests :-) As parallel test execution is extended inside of fixtures, I'm wondering how that would apply to how I'm currently using NUnit with Selenium. I have a base test class that contains a few fields, the most important of which are: a Browser object that encapsulates my WebDriver instance, a pair of Pages objects that encapsulate properties I use to access Selenium Page Objects (bound to the Browser), and a UserAccount object that represents a user account in the system (I build up and tear down user accounts on the fly for each test). In my [SetUp] method I instantiate those objects to make them available to each [Test], the tests use the objects, and then the [TearDown] determines if the test failed, dumps logs/screenshot if so, and calls into that UserAccount instance again to delete the UserAccount from the system.

As I pick up more knowledge and read more in this repo it sounds like the direction is to NOT create a new instance of the test class for each [Test] because each [Test] should be able to stand alone. What I'm using [SetUp] and [TearDown] for at the moment is to avoid code duplication - not doing so would add around 6000 lines of code to my test assembly and also make maintenance a nightmare any time I wanted to change how I tear down the Selenium tests or handle logging/screenshots of failures for all ~150 tests (and growing). What would be the proper way to approach in-fixture parallel test execution in this environment, or should I be waiting until there is the 'new type of TestFixture' you mentioned? I appreciate your insight!

**Comment by CharliePoole** [comment]
NUnit broke with the "tradition" established by junit and sunit in using the same instance for each test. That happened around the year 2000 or a bit before. So there isn't really a "direction" here - it's more a matter of the status quo. Over the years, there have been people who called this a bad design decision. There are definitely pros and cons but overall it seems to have held up.

So, the question arises... has anything changed now that would cause us to reverse that decision. Parallel test execution definitely breaks a scenario like yours. It seems to me there are three generic solutions to the problem, which could be stated as - 
 1. Don't write your tests that way!
 2. Don't use test case parallelism.
 3. Modify how NUnit works.
The first two options would be in your hands, of course.

Option 1 means essentially that you can't save anything in instance state as part of an individual test. SetUp counts as part of an individual test, of course. Of course, what you are doing is pretty much the standard way to use NUnit, so that's a pretty big impact on users. Some people may be able to change, however, using directly called methods on the one hand or OneTimeSetUp on the other.

Option 2 seems like a pretty easy one for users, on the other hand. If the structure of your tests is not consistent with having test cases run in parallel, then don't do it! It's an option after all and one that does not necessarily make tests run any faster than fixture-level parallelism depending on the nature of the tests themselves.

The third option is what we are discussing here. @

**Comment by CharliePoole** [comment]
NUnit broke with the "tradition" established by junit and sunit in using the same instance for each test. That happened around the year 2000 or a bit before. So there isn't really a "direction" here - it's more a matter of the status quo. Over the years, there have been people who called this a bad design decision. There are definitely pros and cons but overall it seems to have held up.

So, the question arises... has anything changed now that would cause us to reverse that decision. Parallel test execution definitely breaks a scenario like yours. It seems to me there are three generic solutions to the problem, which could be stated as - 
 1. Don't write your tests that way!
 2. Don't use test case parallelism.
 3. Modify how NUnit works.
The first two options would be in your hands, of course.

Option 1 means essentially that you can't save anything in instance state as part of an individual test. SetUp counts as part of an individual test, of course. Of course, what you are doing is pretty much the standard way to use NUnit, so that's a pretty big impact on users. Some people may be able to change, however, using directly called methods on the one hand or OneTimeSetUp on the other.

Option 2 seems like a pretty easy one for users, on the other hand. If the structure of your tests is not consistent with having test cases run in parallel, then don't do it! It's an option after all and one that does not necessarily make tests run any faster than fixture-level parallelism depending on the nature of the tests themselves.

The third option is what we are discussing here. @chris-smith-zocdoc requested that parallelism at the test fixture level be tied in with a change in the lifetime of test objects. I object to the tie-in. There are many more kinds of tests than browser tests and even those are not done the same way by everyone. Long experience with NUnit has taught us that features need to be as orthogonal as possible so that users can combine them in the way they desire.

Whether we implement the feature at all, depends on feedback from both our users and team members. It has only just been requested, so let's wait and see. I will say that it is a fairly complex change, so it has to give a pretty significant value to a majority of users if it's going to go anywhere. I intended the idea of embedding it in a separate fixture type as a way to isolate the change so we don't break tests of users who have no need for the change.

You'll have to ask yourself whether you need the change. It seems to me that your simplest course is to continue to run fixtures, rather than tests in parallel. In cases where you have a lot of fixtures and a small number of tests per fixture, test case parallelism doesn't seem to be worth the trouble.

Off topic... Everyone is inexperienced at something! 

I do have a particular dislike of companies who segregate testing from development and assign either the new guys or those they have less confidence in to testing. It's an approach that demeans both the people and the importance of testing. 😠 Anyway, that's what I was referring to.

**Comment by Suremaker** [comment]
Hello,

I'd like to back the @chris-smith-zocdoc request of implementing parallel test method execution in a thread safe manner.

To rationalize it, I'd like to put an argument that the nature of tests (especially unit tests) is that given piece of code can be tested in isolation including fact that tests should not interfere with each others. The non isolated tests will cause nondeterministic results ending up with unstable tests.

The parallel test execution implemented so far in nunit was well supporting the concept of executing tests in isolation, as the most granular mode for running tests was fixture level. This means that it is relatively easy to make tests projects running in parallel.

Enabling ability of running the test methods with parallel while fixture instance is shared between tests will violate the rule of performing tests in isolation. While I agree with @CharliePoole that it is not possible to make a framework that will make impossible for users to make errors, I think that framework should not allow to make errors easily or unintentionally.

There is a comment above:
> Traditionally, the NUnit view has been that we can use the same instance for multiple tests, as well as instance methods for tests, setup and teardown, because we trust the user not to change state - or only to do it in a very controlled way, usually in SetUp and TearDown.

I could make an argument that [SetUp] and [TearDown] are the places in nunit where users change the test state in the controlled way. One example for it is the setup / teardown of the Selenium driver, mentioned by @tparikka, and generally such Setup / Teardown methods are quite common for higher level tests, like service, acceptance or behavioral ones that usually have more complex setup. However it is also quite frequent that Setup methods are used for setting up stuff like mocks for unit tests.

Of course there may be an argument that using [SetUp]/[TearDown] is a bad practice and all code related to test should be in the test method, but at the end this is the capability that nunit offers from many years, and for the pragmatic reasons (like one described by @tparikka) is used often. Now with the instance sharing during parallel test execution, those methods will simply not work as expected.

Just to wrap up the thought on that comment, I do not think this is about trusting users to do the right things because at the end everybody wants to write the best and most solid tests, but rather making a framework to prevent user from doing silly mistakes. For example, please just take a look at the tests from nunit repo. As an exercise I have set `[assembly: Parallelizable(ParallelScope.All)]` to all test projects. When I tried to run tests from console, the runner hung with a deadlock, but Resharper managed to report quite a few failing tests before it locked up.

Below there are example classes with failing tests:
* [ConstraintTestBase.cs#L31](https://github.com/nunit/nunit/blob/f5a8ae729ae7010e88d549a88580c183fbe48f5c/src/NUnitFramework/tests/Constraints/ConstraintTestBase.cs#L31)
* [ComparisonConstraintTestBase.cs#L35](https://github.com/nunit/nunit/blob/65f0b974cb65a6588a54de96f0ce929f57e39f59/src/NUnitFramework/tests/Constraints/ComparisonConstraintTestBase.cs#L35)
* [GreaterThanConstraintTests.cs#L34](https://github.com/nunit/nunit/blob/65f0b974cb65a6588a54de96f0ce929f57e39f59/src/NUnitFramework/tests/Constraints/GreaterThanConstraintTests.cs#L34)

Those are the base classes for wider set of test fixtures, both containing fields:

```c#
    public abstract class ConstraintTestBaseNoData
    {
        protected Constraint theConstraint;
        protected string expectedDescription = "<NOT SET>";
        protected string stringRepresentation = "<NOT SET>";

        [Test]
        public void ProvidesProperDescription()
        {
            Assert.That(theConstraint.Description, Is.EqualTo(expectedDescription));
        }
        /* ... */
    }

   /* ... */

    public abstract class ComparisonConstraintTestBase : ConstraintTestBase
    {
        protected ComparisonConstraint comparisonConstraint;

        [TestCase(null)]
        [TestCase("xxx")]
        public void InvalidDataThrowsArgumentException(object data)
        {
            Assert.Throws<ArgumentException>(() => theConstraint.ApplyTo(data));
        }
        /* ... */
    }

    [TestFixture]
    public class GreaterThanConstraintTests : ComparisonConstraintTestBase
    {
        [SetUp]
        public void SetUp()
        {
            theConstraint = comparisonConstraint = new GreaterThanConstraint(5);
            expectedDescription = "greater than 5";
            stringRepresentation = "<greaterthan 5>";
        }
        /* ... */
   }
```

When I ran those tests with `[assembly: Parallelizable(ParallelScope.All)]` they rendered unpredictable results, while they would be totally fine to run if the test methods were not sharing fixture instance.

---

I have worked in the past on project where we have been using MbUnit framework, and we had thousands of various test cases. Because of the size of project, it was quite important to make those tests running in parallel, but similarly to the plans in nunit, the mbunit was running tests methods in parallel with shared fixture instance.
In order to overcome the problem with race conditions the tests were build in a way that all the fields (like those from nunit tests example) had to be of `ThreadLocal<T>` type (those days it would be rather `AsyncLocal<T>`) and required additional code to properly manage it, making tests even less readable and more difficult to work with.

---

To summarize the long comment (sorry for that), I love the idea of an ability for running test methods in parallel, but mixed with the approach of shared fixture instance will be a root of headache and need of additional unnecessary work (a boilerplate code that for example in xunit is not required to make the same test passing).
Because of that I would strongly vote for implementing it in thread safe manner or at least giving users an ability to control if Fixture instance should be shared per test or not.

**Comment by CharliePoole** [comment]
@Suremaker I don't want to change this into a dialog between me and the advocates for this feature, so I won't respond on every point. I"ve asked for other team members to give their opinion, and I think it's really better to see what the others think before trying to resolve this.

That said, I want to tell you that the theoretical basis of your argument is quite correct, in my view. If we were designing a unit test framework, I we would hash these things out and I would lean your way, particularly from the point of view of a micro-test based TDD. That, however, is not what we are doing. Instead we are discussing a proposal to change a fairly fundamental design decision made 17 years ago and rewrite a substantial part of the NUnit framework. Neither you nor I made those decisions but many people have come to rely on them and we can't change them without breaking tests for someone.

Regarding "thread-safe parallelism" it's a joint job for NUnit and the test-writer to implement thread-safety. We can do nothing that guarantees it although we can make it possible and we do. It's just that we may be requiring more work on the user side than we would like.

BTW, regarding the NUnit tests, I have been running the same experiment as you did for the last three years. That was how I concluded that fixture-level parallelism would be much easier for users to achieve in their tests. Now we are focusing on the part of the feature that will necessarily cause more effort for them.

I'm not sure which parts of my stated conclusions you are disagreeing with, so I'll restate them and I hope you will clarify...

1. I don't think we can or should change the entire underlying structure of how NUnit runs tests.
2. I do think we can add a feature that allows running with unshared instances.
3. I don't think that feature should be made mandatory or automatic when running in parallel because some people already have stateless fixtures and are already running parallel tests successfully.

In sum, if you could go back and make your arguments in the year 2000, I think the guys who are writing  NUnit should listen to you. But we don't have a time machine. :smile:

**Comment by ChrisMaddock** [comment]
I agree this would be a useful option, but shouldn't be the default.

Does it need to be tied into the parallelisation stuff, or could 'class instance per test' be a separate, more general setting?

Sadly, I've long since resigned to the fact that the test suites I work with will never run test cases in parallel. I guess we're all just those less-experienced-programmers!

**Comment by Suremaker** [comment]
I definitely agree that changing the behaviour set 17 years ago is not right way to go and it would be very confusing, but having said that the option 2 from @CharliePoole response sounds viable.

I also agree it would be probably a better idea to separate this `class instance per test` feature from parallel execution. At the end, it may be used in more broader context like to simplify the test classes in a way where all fields could be then initialized at declaration and not have to be reinitialized during Setup.

**Comment by CharliePoole** [comment]
Marked as design so we can figure out how this will be exposed to users before we start working on the code.

**Comment by chris-smith-zocdoc** [comment]
To give us a starting point for discussion, here's my proposal:

`[InstanceLifetime]` is new attribute that can be applied at the class level.  The constructor takes a `Lifetime` enum that controls the instances lifetime which is either `Lifetime.SingleInstance` (the default) or `Lifetime.InstancePerTest`.  Two additional attributes can be added as shorthand for these, `[SingleInstance]` and `[InstancePerTest]`

Example usages

```cs
[InstanceLifetime(Lifetime.InstancePerTest)]
class MyTests { ... }

[InstanceLifetime(Lifetime.SingleInstance)]
class MyTests { ... }

[InstancePerTest]
class MyTests { ... }

[SingleInstance]
class MyTests { ... }
```

**Comment by jnm2** [comment]
`InstancePerTestMethod` or `InstancePerTestCase`?

**Comment by CharliePoole** [comment]
@chris-smith-zocdoc Except your attribute appears on the class - which is a test. I understood @jnm2 to be asking whether you intended your attribute to cause creation of an instance per metnod or an instance per test case. A method marked with [Test] generates a single test case. A method with [TestCase] attributes generates one test case per attribute.

If I understand you correctly, you would like each test case to be run within it's own instance.

**Comment by chris-smith-zocdoc** [comment]
Ah yes, that was my intention.

**Comment by CharliePoole** [comment]
You might consider creating a table of the execution order under the existing approach versus the instance per test case approach. It would need to include SetUp, TearDown, the OneTime variants and any ActionAttributes.

**Comment by chris-smith-zocdoc** [comment]
This is my current understanding of the execution order from reading though SimpleWorkItem and CompositeWorkItem. Please correct anything I got wrong. 


Current reverse chronological order

- TestMethod
- IWrapTestMethod (ExpectedException)
- ITestAction (FixtureNotNullTestAttribute )
- Setup + Teardown
- ITestAction (upstream actions)
- IWrapSetUpTearDown (RetryAttribute, MaxTimeAttribute)
- IApplyToContext (DefaultFloatingPointToleranceAttribute, SetCultureAttribute)
- TimeoutCommand
- // Fixture 
- ITestAction
- Setup + Teardown
- ConstructFixtureCommand
- IApplyToContext



Proposed change
- TestMethod
- IWrapTestMethod (ExpectedException)
- ITestAction (FixtureNotNullTestAttribute )
- Setup + Teardown
- ITestAction (upstream actions)
- IWrapSetUpTearDown (RetryAttribute, MaxTimeAttribute)
- **NEW: ConstructFixtureCommand (or similar)**
- IApplyToContext (DefaultFloatingPointToleranceAttribute, SetCultureAttribute)
- TimeoutCommand
- // Fixture 
- ITestAction
- Setup + Teardown
- ConstructFixtureCommand
- IApplyToContext


The reason I placed the fixture creation there was because all the attributed that execute after that expect the fixture to exist/have access to it.

**Comment by andrewlaser** [comment]
+1 vote for implementing this feature.
Using NUnit as a part of our automation framework (WebDriver + C#)
Each thread needs to use a separate driver instance.
Current parallelization mechanizm lets us run only classes (fixtures in parallel).
Would be nice to run tests within class in parallel as well.

**Comment by CharliePoole** [comment]
@andrewlaser I think you aleady understand this, but I need to correct you for the public record. 😄 

NUnit's current mechanism will let you run individual tests in parallel provided your fixture is entirely stateless. That's an important distinction. For example, when I modify the nunit framework tests to run all test cases in parallel (using `ParallelScope.Children` at the assembly level) a few dozen of the tests, out of more than 4000 faill. Because of the kind of testing we do, most of our fixtures are stateless. Because of the kind of testing you do, I suspect most fixtures have to maintain (modifiable) state. That's typical for web testing.

Given enough parallel fixtures to make use of all cores, fixture parallelization may sometimes be just as efficient as test case parallelization.

I agree that there should be a way to run tests in independent fixture instances. I do not see it as a current item for us to work on, rather as something to do with a major release. Possibly 4.0, possibly 5.0.

@rprouse You haven't wanted to keep "future" stuff around cluttering our issue backlog and have closed a number of them. If we follow that logic, we should close this. OTOH, if you could see your way to maintaining a record of what we hope to do in the future, then maybe we want to do something else... like add to a milestone. In fact, we could even close it and add it to a milestone, re-opening when we are ready to work on it.

**Comment by rprouse** [comment]
@CharliePoole I am not even sure we would consider this future. I am going to close as won't fix in 3. We might decide to implement this in the future and we can revisit this then.

**Comment by jnm2** [comment]
We’re basically saying that if you want to parallelize your tests, and ultimately who doesn’t?,
forget about ever making use of `[SetUp]` and `[TearDown]` and mutable fields.

I ran into something today that made me question this, but so far I think I’m still okay with it.
`[SetUp]` and `[TearDown]` seem to be quite the source of conflicting expectations by NUnit users.
I have mostly avoided them myself, preferring static helper methods and using statements.


This is the ‘obvious’ setup, moving the onerous repetitive locals to class fields and herlper methods,
and goes completely wrong in the presence of parallelization:


```c#
class SomeFixture
{
    private Foo1 setupField1;
    private Foo2 setupField2;
    private Foo2 setupField3;
    private Foo3 setupField4;
    private Foo4 setupField5;
    private Foo5 setupField6;
    private Foo6 setupField7;
        
    [SetUp]
    public async Task Initialize()
    {
        // Assign and set up the seven fields
    }
    
    private Task TestHelper1(Foo2 to, ...)
    {
        // Access some of the setupFields
    }
    
    private Task TestHelper2(Foo2 from, Foo2 to, ...)
    {
        // etc
    }
    
    
    [Test]
    public async Task Test1()
    {
        // Operate on many of the setupFields and use TestHelper methods
    }
    
    [Test]
    public async Task Test2()
    {
        // etc etc
    }
    
    // etc
}
```

Thread-safe version:


```c#
static class SomeFixture
{
    private sealed class Common // I need a better name
                                // This class would be IDisposable if it needed teardown
    {
        public Foo1 SetupField1 { get; } // Yay, everything can bow be readonly!
        public Foo2 SetupField2 { get; }
        public Foo2 SetupField3 { get; }
        public Foo3 SetupField4 { get; } = new SomeFoo3(); // Yay, field initializers run per test now!
        public Foo4 SetupField5 { get; }
        private readonly Foo5 setupField6; // Yay, better encapsulation because these are only for TestHelper methods!
        private readonly Foo6 setupField7;
        
        private Common(Foo1 setupField1, Foo2 setupField2, Foo2 setupField3, Foo4 setupField5)
        {
            SetupField1 = setupField1;
            SetupField2 = setupField2;
            SetupField3 = setupField3;
            SetupField5 = setupField5;
            // Assign and set up setupField6 and setupField7
        }
        
        public static async Task<Common> SetUp()
        {
            // Set up locals setupField1, setupField2, setupField3 and setupField5
            return new Common(setupField1, setupField2, setupField3, setupField5);
        }
    
        public Task TestHelper1(Foo2 to, ...)
        {
            // Access some of the setupFields
        }
        
        public Task TestHelper2(Foo2 from, Foo2 to, ...)
        {
            // etc
        }
    }    
    
    [Test]
    public static async Task Test1()
    {
        var c = await Common.SetUp();
        
        // Operate on the c.SetupFields and use c.TestHelper methods
    }
    
    [Test]
    public static async Task Test2()
    {
        var c = await Common.SetUp();
        
        // etc etc
    }
    
    // etc
}
```

Is it worth adding an `[InstancePerTestCase]` feature to the fixture?
The absence of that feature forced me to write cleaner code, so right now I don’t think so.

**Comment by CharliePoole** [comment]
@jnm2 We're both writing about a request that has already been turned down, but I think it's worth a few more words anyway.

Basically I agree with you. SetUp and TearDown are better replaced by their one-time equivalents if test cases are going to be run in parallel. That way, you can't make any mistakes. Essentially, if you can't accomplish what you want using OneTimeSetUp and OneTimeTearDown, then the test cannot be run in parallel. I might be in favor of a feature that either didn't allow SetUp and TearDown or didn't run tests in parallel if they had either one of them. The first would be a better approach for users, but hard to implement since we don't know how the tests will be run when we are loading them. So I think it has to either be ignoring the parallelism or giving a runtime error. And I'm not sure if either can be done outside a major version change.

I do want to comment on your "who doesn't?" remark. Test case parallelism can be very useful. But too many of our users simply assume they need it. If you have four processors and can keep four parallel fixtures running all the time, that's actually more efficient than test case parallelism would be, since there is substantially less overhead. With a bit more effort, it could be optimized even further. Maybe, for most users, we should concentrate on that.

I'm working on a new framework and I'm starting with the idea that a fixture using a separate instance per test is entirely different from one that uses the same instance for all tests. After many years at this, I think instance-per-test is a better choice for micro- or unit- tests. It's just that with NUnit, we are __not__ creating a new test framework and have to look at compatibility. If and when there is a 4.0 release, maybe we could start to make some changes, leading up to a full revision for 5.0.

**Comment by jnm2** [comment]
@CharliePoole 
> If you have four processors and can keep four parallel fixtures running all the time, that's actually more efficient than test case parallelism would be, since there is substantially less overhead.

This is only the case if you are doing a full test run and not just repeatedly running one or two select fixtures while writing code. That was my main pain point and I noticed right away when test-level parallelization came out. It's not unusual for some kinds of fixture to have a test case source generating between ten and a hundred test cases, each case taking anywhere from milliseconds to a minute, and it's not unusual for me to run just that fixture and wait for feedback on it. Integration tests and algorithm tests tend to be this way.

**Comment by CharliePoole** [comment]
@jnm2 Actually, from my experience visiting a lot of teams over the years, I think it is unusual. I understand it's not unusual for you or for other high end dev doing certain kinds of work, but those dev are not the average.

I agree we need to provide the features needed for those high-end cases, but I want to avoid giving the impression that the average shop is somehow deficient if they don't use those features.

**Comment by jnm2** [comment]
I see, good point.

**Comment by jnm2** [comment]
Please see https://github.com/nunit/nunit/issues/2574 for a proposal to add an instance-per-test-case-feature.
